### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by crsegerie**

### CODE REVIEW

File: commands/predict.py [hunk 1]
```diff
@@ -69,7 +69,14 @@ def main(
     batch_size: int,
     clear_dataset_cache: bool,
 ):
-    """Compute predictions and upload them to Kili."""
+    """Compute predictions and upload them to Kili.
+
+    In order to use this command, you must first use the 'train' command.
+    This command reuses the model that is stored at the end of the training.
```
Comment (PierreLeveau) [inline commands/predict.py:75]: This command uses a model stored after running the `train` command.

File: commands/predict.py [hunk 2]
```diff
@@ -69,7 +69,14 @@ def main(
...
+    We advise you to use this command once in '--dry-run' mode.
```
Comment (PierreLeveau) [inline commands/predict.py:76]: We advise you to use this command once in '--dry-run' mode to preview the predictions.

File: commands/prioritize.py [hunk 3]
```diff
@@ -128,7 +130,9 @@ def __init__(
         Raises:
         ConnectionError: If no available port is found.
         """
-        assert len(embeddings) > 0 and len(embeddings[0]) > 0
+        assert (
```
Comment (PierreLeveau) [inline commands/prioritize.py:133]: can you add a reference of the method used for the prioritization in the docstring?

File: commands/prioritize.py [hunk 4]
```diff
@@ -370,6 +380,17 @@ def main(
     The goal is to find a collection of assets that are most relevant to prioritize.
     We embedded the assets using a generic model, and then use a strategy that is a mixture of
     diversity sampling, uncertainty sampling, and random sampling to sorts the assets.
+
+    diversity_sampling=1
+    This command is useful in the case where many images are very similar.
```
Comment (PierreLeveau) [inline commands/prioritize.py:385]: This command is useful when many images are very similar.

File: commands/prioritize.py [hunk 5]
```diff
@@ -370,6 +380,17 @@ def main(
...
+    This command will cluster the similar images and will first choose a representative of
+    each cluster.
+
+    uncertainty_sampling=1
+    This command is useful in case we want to privilege the labelling of images that are classified
```
Comment (PierreLeveau) [inline commands/prioritize.py:390]: This command is useful when we want to prioritize the labeling of images that are classified

File: commands/train.py [hunk 6]
```diff
@@ -85,6 +85,10 @@ def main(
 
 
     If there are multiple jobs in your projects, a model will be trained on each job.
+    KiliAUtoML will automatically select the right type of model according to your project
```
Comment (PierreLeveau) [inline commands/train.py:88]: Kili AutoML will automatically select the model types that match your Kili project's labeling tasks (...

File: commands/train.py [hunk 7]
```diff
@@ -85,6 +85,10 @@ def main(
...
+    (image classification, text classification, image segmentation)
+    Once the model is trained, the model is stored in the kiliautoml cache.
```
Comment (PierreLeveau) [inline commands/train.py:90]: ~kiliautoml~

File: docs/architecture.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+
+# Architecture
+
+## KiliModels
+
+- kiliautoml/models/_base_model.py : Gives an example of a Parent Model class.
```
Comment (PierreLeveau) [inline docs/architecture.md:6]: put "[kiliautoml/models/_base_model.py]" to get relative links, ` ` if you want to cite them.

File: docs/architecture.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+## The command file
+
+- When using the line command, the instruction is redirected to the file corresponding to the requested command.
+- The documentation and the default parameters of each command are located in the commands/common_args.py file
+- You can read the click librairy documentation, section https://click.palletsprojects.com/en/8.1.x/setuptools/
+
+## The Helper file
+
+The helper file is a file with little structure, but allowing to factorize in the same file functions reused several times in different other files.
+You will find utilities to download asets, to download projects, and other functions with a more rare use.
```
Comment (PierreLeveau) [inline docs/architecture.md:19]: more rarely used.

File: docs/architecture.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+
+## The Typing file
+
+Since Python 3.6, it is possible to use type hint in Python. We use Pylance and Mypy to check the consistency of types. We have finely tuned Pylance to suit our needs.
```
Comment (PierreLeveau) [inline docs/architecture.md:24]: type hints

File: docs/architecture.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+We type AutoML assuming that the executed version is python 3.7, ie, the version of python used by google Colab. When using advanced type objects, we import them from the typing_extension library, which allows us to be backward compatible and use python 3.7 up to 3.10.
+
+
+We took a particular care to type hint the AutoML library in a rigorous way. This makes the development experience more pleasant and increase the speed of iteration considerably.
+
+But we use a relatively complicated type of system:
+- The data structures are either in the form of TypedDict, DataClass or pydantic model.
+    - TypedDict allows to use python dictionaries without overhead, but with the benefit of Pylance type checking. The Kili SDK uses dictionaries, so it is the default solution to type json for assets.
+    - DataClass is used when we don't work directly on assets. Dataclasses are typically obkects specific to autoML, but which don't exist in the SDK documentaiton.
```
Comment (PierreLeveau) [inline docs/architecture.md:33]: objects
documentation

File: docs/architecture.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    - Pydantic is a type verification library. It is mainly used when converting typedDict representing annotations into standardized annotations. Before standardizing the annotations, we proceed to verifications to check that the labels have the right format (example: a bbox must contain 4 points, a polygon at least 3 points, etc...)
```
Comment (PierreLeveau) [inline docs/architecture.md:34]: explain "standardized annotations"

File: docs/architecture.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+- You can check the list of Pylance parameters in the pyproject.toml. For exemple,
+  - reportIncompatibleMethodOverride is very important to ensure that the child classes are consistent with the Parent class.
```
Comment (PierreLeveau) [inline docs/architecture.md:36]: `reportIncompatibleMethodOverride`

File: docs/architecture.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  - reportUnusedVariable is important to ensure that all paramters in a function are used. Is a parameter is not used you need to use something like '_ = unused_parameter'.
```
Comment (PierreLeveau) [inline docs/architecture.md:37]: `reportUnusedVariable`,  parameters

**Review by PierreLeveau — CHANGES_REQUESTED** [review]
I did not finish the review. Can you please make a pass to:
- fix the typos
- code citations should be enclosed with backticks "`"
- file citations should be enclosed with "()[]" to navigate  to them

File: commands/prioritize.py (hunk 3 above)
Comment (crsegerie) [reply commands/prioritize.py:133]: ?

File: docs/architecture.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
+
+# Architecture
+
+## KiliModels
+
+- [kiliautoml/models/_base_model.py] : Gives an example of a parent Model class.
```
Comment (PierreLeveau) [inline docs/architecture.md:6]:
```suggestion
- (`kiliautoml/models/_base_model.py`)[kiliautoml/models/_base_model.py] : Gives an example of a parent Model class.
```
same for all the links.

File: docs/architecture.md [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+- [kiliautoml/models/_detectron2_semantic_segmentation.py] : Gives an example of model (Child class)
+- [kiliautoml/models/kili_auto_model.py] : Only public interface. This class encapsulates all the other models, but be careful, it does not exactly match the model of the parent class in [_base_model.py]
+
+## The command file
+
+- When using the line command, the instruction is redirected to the file corresponding to the requested command.
```
Comment (PierreLeveau) [inline docs/architecture.md:12]: `command line`

File: docs/architecture.md [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+- The documentation and the default parameters of each command are located in the commands/common_args.py file
```
Comment (PierreLeveau) [inline docs/architecture.md:13]: ... (`commands/common_args.py`)[commands/common_args.py] file.

File: docs/architecture.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+But we use a relatively complicated system of type:
+- The data structures are either in the form of TypedDict, DataClass or pydantic model.
+    - TypedDict allows to use python dictionaries without overhead, but with the benefit of Pylance type checking. The Kili SDK uses dictionaries, so it is the default solution to type json containing assets.
+    - DataClass is used when we don't work directly on assets. Dataclasses are typically objects specific to autoML, but which don't exist in the SDK documentation.
+    - Pydantic is a type verification library. It is mainly used when converting typedDict representing annotations into standardized annotations (The standardized annotations is a way to represent any annotation from any modality to compute the label errors). Before standardizing the annotations, we proceed to check that the labels have the right formats (example: a bbox must contain 4 points, a polygon at least 3 points, etc...)
+- You can check the list of Pylance parameters in the [pyproject.toml] file. For exemple,
```
Comment (PierreLeveau) [inline docs/architecture.md:35]: (`pyproject.toml`)[pyproject.toml]

File: docs/architecture.md [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+  - `reportIncompatibleMethodOverride` is very important to ensure that the child classes are consistent with the Parent class.
+  - `reportUnusedVariable` is important to ensure that all paramters in a function are used. If a parameter is not used you need to use something like `_ = unused_parameter`.
+
+
+The types are then distributed to all other files in the application.
+
+## Particularities of each modality
+
+- Image classification: This is the only modality that benefits from the prioritization module. It uses plain pytorchvision in backend.
```
Comment (PierreLeveau) [inline docs/architecture.md:44]: as a backend (and same for the next 2 points)

Comment (PierreLeveau) [inline docs/architecture.md:44]: **Image classification**

Comment (PierreLeveau) [reply docs/architecture.md:44]: (also put in bold fonts the subsequent points)

File: docs/architecture.md [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+- Text classification: It uses hugggingface in backend.
+- Image segmentation and polygon: Image segmentation and bbox are two different things in the Kili interface but share exactly the same code in AutoML. The difference between the two is not a qualitative difference but simply a quantitative one: image segmentation = polygon with many sides, polygon = polygon with relatively few sides. They use the Detectron library in backend.
+
+
+## System used in the LabelError command
+
+- Image classification: Image classification uses cleanlab in the backend to detect annotation errors. The paradigm used by cleanlab consists in training several models on different parts of the data, then comparing the predictions of each model.
+- Object Detection: Polygon, Bbox, Image Segmentation: These three modalities use the same error detection system which is hand coded. Once a model has been trained, we compare the model predictions with the manual annotations.
```
Comment (PierreLeveau) [inline docs/architecture.md:52]: Add a sentence that summarizes the error detection procedure for these tasks.

**Review by PierreLeveau — CHANGES_REQUESTED** [review]
Thanks for the changes, a few are remaining.

**Review by PierreLeveau — APPROVED** [review]

**Pull request merged**